modules/lang-painless/src/main/java/org/elasticsearch/painless/FunctionRef.java

```diff
@@ -165,6 +168,16 @@ public static FunctionRef create(PainlessLookup painlessLookup, FunctionTable fu
 
                 delegateMethodName = painlessMethod.javaMethod.getName();
                 delegateMethodType = painlessMethod.methodType;
+
+                // interfaces that override a method from Object receive the method handle for
+                // Object rather than for the interface; we change the first parameter to match
+                // the interface type so the constant interface method reference is correctly
+                // written to the constant pool
+                if (delegateInvokeType != H_INVOKESTATIC &&
+                        painlessMethod.javaMethod.getDeclaringClass() != painlessMethod.methodType.parameterType(0)) {
+                    delegateMethodType = delegateMethodType.changeParameterType(0, painlessMethod.javaMethod.getDeclaringClass());
```

Comment:
I think we should throw a state exception (or something similar) if the parameterType(0) is not Object. It could also just be an assert, but in essence I'm concerned we could mask another similar problem that we don't understand now.

Reply:
That's a good idea. I added an IllegalStateException if it's not Object.